File: backend/app/models/pedido.py
from datetime import datetime, timezone
from decimal import Decimal
from typing import TYPE_CHECKING, Optional, List

from sqlmodel import Field, Relationship, SQLModel

if TYPE_CHECKING:
    from app.models.detalle_pedido import DetallePedido
    from app.models.direccion_entrega import DireccionEntrega
    from app.models.historial_estado_pedido import HistorialEstadoPedido
    from app.models.usuario import Usuario


class Pedido(SQLModel, table=True):
    __tablename__ = "pedido"

    id: Optional[int] = Field(default=None, primary_key=True)
    usuario_id: int = Field(foreign_key="usuario.id", index=True)
    direccion_id: Optional[int] = Field(default=None, foreign_key="direccion_entrega.id", index=True)
    estado_codigo: str = Field(default="PENDIENTE", foreign_key="estado_pedido.codigo", index=True, max_length=40)
    forma_pago_codigo: str = Field(foreign_key="forma_pago.codigo", index=True, max_length=40)

    subtotal: Decimal = Field(default=0, max_digits=10, decimal_places=2, ge=0)
    descuento: Decimal = Field(default=0, max_digits=10, decimal_places=2, ge=0)
    costo_envio: Decimal = Field(default=0, max_digits=10, decimal_places=2, ge=0)
    total: Decimal = Field(default=0, max_digits=10, decimal_places=2, ge=0)
    notas: Optional[str] = Field(default=None, max_length=500)

    activo: bool = Field(default=True)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    deleted_at: Optional[datetime] = Field(default=None)

    usuario: "Usuario" = Relationship(back_populates="pedidos")
    direccion: "DireccionEntrega" = Relationship(back_populates="pedidos")
    detalles: List["DetallePedido"] = Relationship(back_populates="pedido")
    historial_estados: List["HistorialEstadoPedido"] = Relationship(back_populates="pedido")

File: frontend/src/types/pedido.ts
export interface DetallePedido {
  pedido_id: number;
  producto_id: number;
  cantidad: number;
  nombre_producto_snap: string;
  precio_unitario_snap: number;
  subtotal_snap: number;
  personalizacion?: Record<string, unknown> | null;
  created_at: string;
}

export interface HistorialEstadoPedido {
  id: number;
  pedido_id: number;
  estado_desde?: string | null;
  estado_hacia: string;
  usuario_id?: number | null;
  motivo?: string | null;
  created_at: string;
}

export interface Pedido {
  id: number;
  usuario_id: number;
  direccion_id?: number | null;
  estado_codigo: string;
  forma_pago_codigo: string;
  tipo_entrega: "RETIRO" | "ENVIO" | string;
  domicilio_retiro_snap?: string | null;
  datos_transferencia_snap?: string | null;
  subtotal: number;
  descuento: number;
  costo_envio: number;
  total: number;
  notas?: string | null;
  activo: boolean;
  created_at: string;
  updated_at: string;
  deleted_at?: string | null;
  detalles: DetallePedido[];
  historial_estados: HistorialEstadoPedido[];
}

export interface PedidoEstadoUpdatePayload {
  estado_codigo: string;
  motivo?: string | null;
  recuperar_stock?: boolean;
}

export interface EstadoPedido {
  id: number;
  codigo: string;
  nombre: string;
  orden: number;
  permite_cancelacion_cliente: boolean;
  es_final: boolean;
  activo: boolean;
}

export interface FormaPago {
  id: number;
  codigo: string;
  nombre: string;
  descripcion?: string | null;
  activo: boolean;
}

export interface PedidoItemCreate {
  producto_id: number;
  cantidad: number;
  personalizacion?: Record<string, unknown> | null;
}

export interface PedidoCreatePayload {
  direccion_id?: number | null;
  tipo_entrega?: "RETIRO" | "ENVIO" | string;
  forma_pago_codigo: string;
  notas?: string | null;
  descuento: number;
  costo_envio: number;
  items: PedidoItemCreate[];
}
